grpc/src/main/java/com/linecorp/armeria/server/grpc/HttpJsonTranscodingService.java

```diff
@@ -571,10 +578,28 @@ private HttpResponse serve0(ServiceRequestContext ctx, HttpRequest req,
                 } else {
                     try {
                         ctx.setAttr(FramedGrpcService.RESOLVED_GRPC_METHOD, spec.method);
-                        // Set JSON media type (https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter#sending-arbitrary-content)
+                        final MediaType mediaType;
+                        final HttpData content;
+
+                        // https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_json_transcoder_filter#sending-arbitrary-content
+                        if (HttpBody.getDescriptor().equals(spec.methodDescriptor.getInputType())) {
+                            // TODO: Determine media type dynamically from content.
+                            mediaType = ctx.request().contentType();
+                            content = convertToArbitraryHttpData(ctx, clientRequest);
+                        } else {
+                            mediaType = MediaType.JSON_UTF_8;
+                            content = convertToJson(ctx, clientRequest, spec);
+                        }
+
+                        // Media type can only be nullif the request doesn't have a content type.
+                        if (mediaType == null) {
+                            throw new IllegalArgumentException("Unsupported content-type: " +
+                                                               clientRequest.contentType());
+                        }
+
                         frameAndServe(unwrap(), ctx, grpcHeaders.build(),
-                                      convertToJson(ctx, clientRequest, spec), responseFuture,
-                                      generateResponseBodyConverter(spec), MediaType.JSON_UTF_8);
+                                      content, responseFuture,
+                                      generateResponseBodyConverter(spec), mediaType);
```

Comment:
Here it is trickier, we should pass the media type dynamically before responding. It is going to involve some minor changes to the deframe logic. (Such as determining the content type dynamically after response is generated and falling back if response is not generated).